benchmarks/performance-poetry/complex-poetry/src/androidTest/java/com/squareup/benchmarks/performance/complex/poetry/RenderPassTest.kt

```diff
@@ -0,0 +1,126 @@
+import com.squareup.benchmarks.performance.complex.poetry.robots.waitForPoetry
+import org.junit.Assert.fail
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+
+/**
+ * Test that is used to confirm that the number of render passes and the fresh rendering ratio
+ * is constant.
+ */
+@RunWith(AndroidJUnit4::class)
+class RenderPassTest {
+  data class Scenario(
+    val title: String,
+    val config: SimulatedPerfConfig,
+    val expectedPasses: Int,
+    val expectedRatio: Double
+  )
+
+  private val renderPassCountingInterceptor = RenderPassCountingInterceptor()
+  private val device = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
+  private lateinit var context: Context
+
+  @Before fun setup() {
+    context = ApplicationProvider.getApplicationContext()
+    PerformancePoetryActivity.installedInterceptor = renderPassCountingInterceptor
+  }
+
+  @Test fun renderPassCounterComplexWithInitializingState() {
+    runRenderPassCounter(COMPLEX_INITIALIZING)
+  }
+
+  @Test fun renderPassCounterComplexNoInitializingState() {
+    runRenderPassCounter(COMPLEX_NO_INITIALIZING)
+  }
+
+  private fun runRenderPassCounter(scenario: Scenario) {
+    val intent = Intent(context, PerformancePoetryActivity::class.java).apply {
+      addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
+      addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK)
+      putExtra(PERF_CONFIG_EXTRA, scenario.config)
+    }
+    device.wakeUp()
+    device.pressHome()
+    // Do these in landscape so we have both the 'index' and the 'detail' showing.
+    device.landscapeOrientation()
+
+    InstrumentationRegistry.getInstrumentation().context.startActivity(intent)
+
+    device.waitForPoetry()
+    device.waitForIdle()
+
+    device.openRavenAndNavigate()
+
+    val totalRenderPasses = renderPassCountingInterceptor.renderEfficiencyTracking.totalRenderPasses
+    val renderPassSubject = "the number of Render Passes (lower better)"
+    if (totalRenderPasses > scenario.expectedPasses) {
+      uhOh(renderPassSubject, "$totalRenderPasses", scenario.title)
+    } else if (totalRenderPasses < scenario.expectedPasses) {
+      congrats(renderPassSubject, "$totalRenderPasses", scenario.title)
+    }
+
+    val freshRatio = renderPassCountingInterceptor.renderEfficiencyTracking.freshRenderingRatio
+    val ratioSubject = "the fresh rendering ratio (higher better)"
+    if (freshRatio > scenario.expectedRatio) {
+      congrats(ratioSubject, "$freshRatio", scenario.title)
+    } else if (freshRatio < scenario.expectedRatio) {
+      uhOh(ratioSubject, "$freshRatio", scenario.title)
+    }
+  }
+
+  companion object {
+    val COMPLEX_INITIALIZING = Scenario(
+      title = "the 'Raven navigation with initializing state scenario'",
+      config = SimulatedPerfConfig(
+        isComplex = true,
+        useInitializingState = true,
+        complexityDelay = 100L
+      ),
+      expectedPasses = 61,
+      expectedRatio = 0.11891891891891893
```

Comment:
Wondering if you'd want to store the ratio as 2 numbers instead? idk feels weird to store a truncated number.

Reply:
In fact, if you stored 2 numbers, you could then assert against each, and report whether there are less things fresh or more overall things.